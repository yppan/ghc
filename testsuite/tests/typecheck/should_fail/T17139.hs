{-# LANGUAGE GADTs #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE AllowAmbiguousTypes #-}

module T17139 where

import T17139a

type family TypeFam f fun where
  TypeFam f (a -> b) = f a -> TypeFam f b

lift :: (a -> b) -> TypeFam f (a -> b)
lift f = \x -> _ (f <*> x)
