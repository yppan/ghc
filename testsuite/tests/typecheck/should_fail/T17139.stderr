
T17139.hs:15:10: error:
    • Couldn't match type ‘f’ with ‘(->) a’
      Expected: TypeFam f (a -> b)
        Actual: (a -> a) -> f a -> TypeFam f b0
      ‘f’ is a rigid type variable bound by
        the type signature for:
          lift :: forall a b (f :: * -> *). (a -> b) -> TypeFam f (a -> b)
        at T17139.hs:14:1-38
    • The lambda expression ‘\ x -> _ (f <*> x)’
      has one value argument,
        but its type ‘TypeFam f (a -> b)’ has none
      In the expression: \ x -> _ (f <*> x)
      In an equation for ‘lift’: lift f = \ x -> _ (f <*> x)
    • Relevant bindings include
        f :: a -> b (bound at T17139.hs:15:6)
        lift :: (a -> b) -> TypeFam f (a -> b) (bound at T17139.hs:15:1)

T17139.hs:15:16: error:
    • Found hole: _ :: (a -> b0) -> f a -> TypeFam f b0
      Where: ‘b0’ is an ambiguous type variable
             ‘a’, ‘f’ are rigid type variables bound by
               the type signature for:
                 lift :: forall a b (f :: * -> *). (a -> b) -> TypeFam f (a -> b)
               at T17139.hs:14:1-38
    • In the expression: _ (f <*> x)
      The lambda expression ‘\ x -> _ (f <*> x)’ has one value argument,
        but its type ‘TypeFam f (a -> b)’ has none
      In the expression: \ x -> _ (f <*> x)
    • Relevant bindings include
        x :: a -> a (bound at T17139.hs:15:11)
        f :: a -> b (bound at T17139.hs:15:6)
        lift :: (a -> b) -> TypeFam f (a -> b) (bound at T17139.hs:15:1)

T17139.hs:15:19: error:
    • Couldn't match type ‘b’ with ‘a -> b0’
      Expected: a -> a -> b0
        Actual: a -> b
      ‘b’ is a rigid type variable bound by
        the type signature for:
          lift :: forall a b (f :: * -> *). (a -> b) -> TypeFam f (a -> b)
        at T17139.hs:14:1-38
    • In the first argument of ‘(<*>)’, namely ‘f’
      In the first argument of ‘_’, namely ‘(f <*> x)’
      In the expression: _ (f <*> x)
    • Relevant bindings include
        x :: a -> a (bound at T17139.hs:15:11)
        f :: a -> b (bound at T17139.hs:15:6)
        lift :: (a -> b) -> TypeFam f (a -> b) (bound at T17139.hs:15:1)
